command/connect/redirecttraffic/redirect_traffic.go

```diff
@@ -132,23 +148,72 @@ func (c *cmd) generateConfigFromFlags() (iptables.Config, error) {
 			return iptables.Config{}, fmt.Errorf("service %s is not a proxy service", c.proxyID)
 		}
 
-		cfg.ProxyInboundPort = svc.Port
+		// Decode proxy's opaque config so that we can use it later to configure
+		// traffic redirection with iptables.
 		var trCfg trafficRedirectProxyConfig
 		if err := mapstructure.WeakDecode(svc.Proxy.Config, &trCfg); err != nil {
 			return iptables.Config{}, fmt.Errorf("failed parsing Proxy.Config: %s", err)
 		}
 
+		// Set the proxy's inbound port.
+		cfg.ProxyInboundPort = svc.Port
 		if trCfg.BindPort != 0 {
 			cfg.ProxyInboundPort = trCfg.BindPort
 		}
 
-		// todo: Change once it's configurable
+		// Set the proxy's outbound port.
 		cfg.ProxyOutboundPort = iptables.DefaultTProxyOutboundPort
+		if svc.Proxy.TransparentProxy.OutboundListenerPort != 0 {
+			cfg.ProxyOutboundPort = svc.Proxy.TransparentProxy.OutboundListenerPort
+		}
+
+		// Exclude envoy_prometheus_bind_addr port from inbound redirection rules.
+		if trCfg.PrometheusBindAddr != "" {
+			_, port, err := net.SplitHostPort(trCfg.PrometheusBindAddr)
+			if err != nil {
+				return iptables.Config{}, fmt.Errorf("failed parsing host and port from envoy_prometheus_bind_addr: %s", err)
+			}
+
+			cfg.ExcludeInboundPorts = append(cfg.ExcludeInboundPorts, port)
+		}
+
+		// Exclude envoy_stats_bind_addr port from inbound redirection rules.
+		if trCfg.StatsBindAddr != "" {
+			_, port, err := net.SplitHostPort(trCfg.StatsBindAddr)
+			if err != nil {
+				return iptables.Config{}, fmt.Errorf("failed parsing host and port from envoy_stats_bind_addr: %s", err)
+			}
+
+			cfg.ExcludeInboundPorts = append(cfg.ExcludeInboundPorts, port)
+		}
+
+		// Exclude the ListenerPort from Expose configs from inbound traffic redirection.
+		for _, exposePath := range svc.Proxy.Expose.Paths {
```

Comment:
One thing I'm realizing is missing is this bit of magic:
https://www.consul.io/docs/connect/registration/service-registration#checks

If someone enables this, Consul will dynamically expose checks on a listener for each check:
https://github.com/hashicorp/consul/blob/c343544fef094a505d22cf60bd79635cf8e2553e/agent/agent.go#L3839

We don't expose what port is allocated in any of our APIs, similarly to how we don't expose check definitions via any of our APIs, only check statuses. See this comment for more information on why: https://github.com/hashicorp/consul/issues/7330#issuecomment-592015038

There are a few potential solutions to discover those ports off the top of my head:

1. Expand the `structs.HealthCheck` type to include the exposed path.
2. Add a Meta map to checks, like we have for services, and include the exposed path there.
3. Include all check definition details in `structs.HealthCheck`. That comes with the security risk mentioned above.

Reply:
So based on the discussion today we should go with #1.

The exposed ports are generated in a few places. 

One is where we add checks:
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L2610
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L2541

In `addCheck` it's easy to expand the HealthCheck type, since we already have the `check` pointer to it. 

And the other is where we add services with checks:
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L3791
https://github.com/hashicorp/consul/blob/0c9555ec0e2366996cff09d39d1bcb40eaeee435/agent/agent.go#L3818

In these two other methods we'll need to query `a.State.Check`, I think, and modify that.

Reply:
That's a great find! Let me address this in a separate PR so that this one can be included in beta2.